controller/controlplane/controller.go

```diff
@@ -188,65 +116,22 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 			}, nil
 		}
 
-		log.Trace(logger, "controlplane marked for deletion, removing owned cluster roles, cluster role bindings and validating webhook configurations")
-
-		newControlPlane := cp.DeepCopy()
-
-		// ensure that the ValidatingWebhookConfigurations which was created for the ControlPlane is deleted
-		deletions, err := r.ensureOwnedValidatingWebhookConfigurationDeleted(ctx, cp)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		if deletions {
-			log.Debug(logger, "ValidatingWebhookConfiguration deleted")
-			return ctrl.Result{}, nil // ValidatingWebhookConfiguration deletion will requeue
-		}
-
-		// now that ValidatingWebhookConfigurations are cleaned up, remove the relevant finalizer
-		if controllerutil.RemoveFinalizer(newControlPlane, string(ControlPlaneFinalizerCleanupValidatingWebhookConfiguration)) {
-			if err := r.Patch(ctx, newControlPlane, client.MergeFrom(cp)); err != nil {
-				return ctrl.Result{}, err
+		if err := r.InstancesManager.StopInstance(mgrID); err != nil {
+			if errors.As(err, &multiinstance.InstanceNotFoundError{}) {
+				log.Debug(logger, "control plane instance not found, skipping cleanup")
+			} else {
+				return ctrl.Result{}, fmt.Errorf("failed to stop instance: %w", err)
 			}
-			log.Debug(logger, "ValidatingWebhookConfigurations finalizer removed")
-			return ctrl.Result{}, nil // ControlPlane update will requeue
 		}
 
-		// ensure that the clusterrolebindings which were created for the ControlPlane are deleted
-		deletions, err = r.ensureOwnedClusterRoleBindingsDeleted(ctx, cp)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		if deletions {
-			log.Debug(logger, "clusterRoleBinding deleted")
-			return ctrl.Result{}, nil // ClusterRoleBinding deletion will requeue
-		}
-
-		// now that ClusterRoleBindings are cleaned up, remove the relevant finalizer
-		if controllerutil.RemoveFinalizer(newControlPlane, string(ControlPlaneFinalizerCleanupClusterRoleBinding)) {
-			if err := r.Patch(ctx, newControlPlane, client.MergeFrom(cp)); err != nil {
-				return ctrl.Result{}, err
-			}
-			log.Debug(logger, "clusterRoleBinding finalizer removed")
-			return ctrl.Result{}, nil // ControlPlane update will requeue
-		}
-
-		// ensure that the clusterroles created for the controlplane are deleted
-		deletions, err = r.ensureOwnedClusterRolesDeleted(ctx, cp)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		if deletions {
-			log.Debug(logger, "clusterRole deleted")
-			return ctrl.Result{}, nil // ClusterRole deletion will requeue
-		}
-
-		// now that ClusterRoles are cleaned up, remove the relevant finalizer
-		if controllerutil.RemoveFinalizer(newControlPlane, string(ControlPlaneFinalizerCleanupClusterRole)) {
-			if err := r.Patch(ctx, newControlPlane, client.MergeFrom(cp)); err != nil {
-				return ctrl.Result{}, err
+		// remove finalizer
+		controllerutil.RemoveFinalizer(cp, string(ControlPlaneFinalizerCPInstanceTeardown))
```

Comment:
We can use the return value from `RemoveFinalizer` and only update when it returns `true`. This would save some updates when they would not be necessary.